src/net/atlas/db.rs

```diff
@@ -0,0 +1,379 @@
+            }
+        } else {
+            // can just open
+            if readwrite {
+                OpenFlags::SQLITE_OPEN_READ_WRITE
+            } else {
+                OpenFlags::SQLITE_OPEN_READ_ONLY
+            }
+        };
+        let conn =
+            Connection::open_with_flags(path, open_flags).map_err(|e| db_error::SqliteError(e))?;
+
+        let mut db = AtlasDB {
+            conn: conn,
+            readwrite: readwrite,
+        };
+        if create_flag {
+            db.instantiate()?;
+        }
+        Ok(db)
+    }
+
+    // Open a burn database in memory (used for testing)
+    #[cfg(test)]
+    pub fn connect_memory() -> Result<AtlasDB, db_error> {
+        let conn = Connection::open_in_memory().map_err(|e| db_error::SqliteError(e))?;
+
+        let mut db = AtlasDB {
+            conn: conn,
+            readwrite: true,
+        };
+
+        db.instantiate()?;
+        Ok(db)
+    }
+
+    pub fn conn(&self) -> &Connection {
+        &self.conn
+    }
+
+    pub fn tx_begin<'a>(&'a mut self) -> Result<Transaction<'a>, db_error> {
+        if !self.readwrite {
+            return Err(db_error::ReadOnly);
+        }
+
+        let tx = tx_begin_immediate(&mut self.conn)?;
+        Ok(tx)
+    }
+
+    pub fn get_minmax_heights_window_for_page_index(
+        &self,
+        oldest_page_index: u32,
+        newest_page_index: u32,
+    ) -> Result<(u64, u64), db_error> {
+        let qry = "SELECT MIN(block_height) as min, MAX(block_height) as max FROM attachment_instances WHERE page_index >= ?1 AND page_index <= ?2".to_string();
+        let args = [
+            &oldest_page_index as &dyn ToSql,
+            &newest_page_index as &dyn ToSql,
+        ];
+        let mut stmt = self.conn.prepare(&qry)?;
+        let mut rows = stmt.query(&args)?;
+        match rows.next() {
+            Some(Ok(row)) => {
+                let min: i64 = row.get_checked("min")?;
+                let max: i64 = row.get_checked("max")?;
+                Ok((min as u64, max as u64))
+            }
+            _ => Err(db_error::NotFoundError),
+        }
+    }
+
+    pub fn get_attachments_available_at_pages_indexes(
+        &self,
+        pages_indexes: &Vec<u32>,
+        blocks_ids: &Vec<StacksBlockId>,
+    ) -> Result<Vec<Vec<u8>>, db_error> {
+        let mut pages = vec![];
+        for page_index in pages_indexes {
+            let page = self.get_attachments_missing_at_page_index(*page_index, blocks_ids)?;
+            let mut bit_vector = vec![];
+            for (_index, is_attachment_missing) in page.iter().enumerate() {
+                // todo(ludo): use a bitvector instead
+                bit_vector.push(if *is_attachment_missing { 0 } else { 1 });
+            }
+            pages.push(bit_vector);
+        }
+        Ok(pages)
+    }
+
+    pub fn get_attachments_missing_at_page_index(
+        &self,
+        page_index: u32,
+        blocks_ids: &Vec<StacksBlockId>,
```

Comment:
How big can this array get?

Reply:
Good question. It will depend on how good we are at setting the `ATTACHMENTS_INV_PAGE_SIZE` constant in the BNS contract. 
If this value is too low (compared to the number of names being registered / updated per block) then a node requesting attachments for a given block will be requesting multiple inventory pages.
If this value is too high, then a node will be requesting one inventory page, that will be also be including attachments coming from parent block(s).
We can probably use the traffic on v1 for setting this value? This todo is tracked here https://github.com/blockstack/stacks-blockchain/issues/2083.

Reply:
If we dropped `ATTACHMENTS_INV_PAGE_SIZE` from the BNS contract altogether and did all pagination within the Atlas code, then we can take several different steps to control the size of this array.  For example, the Atlas network could have variable page sizes based on how deeply-confirmed each zone file hash is (but we don't have to implement this right now -- because this change would move pagination out of BNS and entirely into Atlas, we can work on it post-launch if need be).

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/commit/2dc4945beffe616add25462b716bf039a3ea8c00